File: kirun-js/src/engine/dsl/lexer/LexerError.ts
import { SourceLocation } from './DSLToken';

/**
 * Lexer error with position information
 */
export class LexerError extends Error {
    constructor(
        message: string,
        public location?: SourceLocation,
        public context?: string,
    ) {
        super(message);
        this.name = 'LexerError';

        // Format error message with location if available
        if (location) {
            this.message = `${message} at ${location}`;
            if (context) {
                this.message += `\n\n${this.formatContext(context, location)}`;
            }
        }
    }

    private formatContext(input: string, location: SourceLocation): string {
        const lines = input.split('\n');
        const lineIndex = location.line - 1;

        if (lineIndex < 0 || lineIndex >= lines.length) {
            return '';
        }

        const line = lines[lineIndex];
        const pointer = ' '.repeat(location.column - 1) + '^';

        return `${line}\n${pointer}`;
    }
}

File: kirun-py/src/kirun_py/dsl/lexer/lexer_error.py
from __future__ import annotations

from typing import Optional

from kirun_py.dsl.lexer.dsl_token import SourceLocation


class LexerError(Exception):
    """Lexer error with position information."""

    def __init__(
        self,
        message: str,
        location: Optional[SourceLocation] = None,
        context: Optional[str] = None,
    ) -> None:
        self.location = location
        self.context = context

        formatted = message
        if location:
            formatted = f"{message} at {location}"
            if context:
                formatted += f"\n\n{self._format_context(context, location)}"

        super().__init__(formatted)

    @staticmethod
    def _format_context(input_text: str, location: SourceLocation) -> str:
        lines = input_text.split('\n')
        line_index = location.line - 1

        if line_index < 0 or line_index >= len(lines):
            return ''

        line = lines[line_index]
        pointer = ' ' * (location.column - 1) + '^'

        return f"{line}\n{pointer}"
